In [1]:
!pip install requests


Test latency for sample data

In [2]:
import requests
import time
import concurrent.futures

# Define the API endpoint (replace with the ngrok public URL)
api_url = "https://26e8-34-145-195-49.ngrok-free.app/qna"  # Update with the actual ngrok URL

# Sample data for testing
sample_data = [
    {
        "context": "The Apollo 11 mission was the first manned mission to land on the Moon.",
        "question": "What was the first manned mission to land on the Moon?"
    },
    {
        "context": "Python is a widely used high-level programming language.",
        "question": "What is Python?"
    },
    # Add more samples as needed
]

# Function to send a request to the QnA API
def send_request(data):
    response = requests.post(api_url, json=data)
    return response.json()

# Measure latency for multiple concurrent requests
def measure_latency_concurrent_requests(sample_data, num_requests=10):
    latencies = []
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(send_request, data) for data in sample_data for _ in range(num_requests)]
        for future in concurrent.futures.as_completed(futures):
            response = future.result()
            latencies.append(response['latency'])
    return latencies

# Run the testing script
latencies = measure_latency_concurrent_requests(sample_data, num_requests=5)
print(f"Average Latency: {sum(latencies) / len(latencies):.2f} seconds")
print(f"All Latencies: {latencies}")


Average Latency: 0.22 seconds
All Latencies: [0.26305103302001953, 0.23623442649841309, 0.2229321002960205, 0.24605917930603027, 0.2501261234283447, 0.2115497589111328, 0.18649649620056152, 0.17915654182434082, 0.18001890182495117, 0.1818091869354248]


Response times (latency) for multiple concurrent requests

In [4]:
# Mount Google Drive to access files
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [10]:
import requests
import time
import concurrent.futures
import pandas as pd

# Define the API endpoint (replace with the actual ngrok URL)
api_url = "https://5fb3-34-145-195-49.ngrok-free.app/qna"  # Update with the actual ngrok URL

# Load dataset from CSV
def load_dataset(file_path):
    try:
        df = pd.read_csv(file_path, encoding='ISO-8859-1')  # Adjust encoding as needed
        print(df.columns)  # Print columns to verify names
        contexts = df['Context'].tolist()  # Use the correct column names
        questions = df['Question'].tolist()
        return [{'context': context, 'question': question} for context, question in zip(contexts, questions)]
    except UnicodeDecodeError as e:
        print(f"Error reading the file: {e}")
        raise
    except KeyError as e:
        print(f"KeyError: {e}. Check the column names in the CSV file.")
        raise

# Function to send a request to the QnA API and measure latency
def send_request(data):
    try:
        start_time = time.time()
        response = requests.post(api_url, json=data)
        end_time = time.time()
        latency = end_time - start_time
        # Print status code and response text for debugging
        print(f"Status Code: {response.status_code}")
        print(f"Response Text: {response.text}")
        response_json = response.json()  # May raise JSONDecodeError if response is not valid JSON
        return latency, response_json
    except requests.exceptions.RequestException as e:
        print(f"Request failed: {e}")
        return None, None
    except ValueError as e:
        print(f"JSON decoding failed: {e}")
        return None, None

# Measure latency for multiple concurrent requests
def measure_latency_concurrent_requests(data, num_requests=10):
    latencies = []
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(send_request, data) for data in data for _ in range(num_requests)]
        for future in concurrent.futures.as_completed(futures):
            latency, _ = future.result()
            if latency is not None:
                latencies.append(latency)
    return latencies

# Update with your dataset file path
file_path = '/content/drive/My Drive/Proplens/qa_data.csv'
sample_data = load_dataset(file_path)

# Run the testing script
latencies = measure_latency_concurrent_requests(sample_data, num_requests=5)
if latencies:
    average_latency = sum(latencies) / len(latencies)
    print(f"Average Latency: {average_latency:.2f} seconds")
    print(f"All Latencies: {latencies}")
else:
    print("No valid latencies recorded.")


Index(['File', 'Context', 'Question', 'Answer'], dtype='object')
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.23428654670715332}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.2130112648010254}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.21226239204406738}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.24605131149291992}
Status Code: 200
Response Text: {"answer":"registration","latency":0.19738316535949707}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.21283888816833496}
Status Code: 200
Response Text: {"answer":"registration","latency":0.19561004638671875}
Status Code: 200
Response Text: {"answer":"registration","latency":0.1901555061340332

n * num_requests =48*10=480

In [11]:
import requests
import time
import concurrent.futures
import pandas as pd

# Define the API endpoint (replace with the actual ngrok URL)
api_url = "https://5fb3-34-145-195-49.ngrok-free.app/qna"  # Update with the actual ngrok URL

# Load dataset from CSV
def load_dataset(file_path):
    try:
        df = pd.read_csv(file_path, encoding='ISO-8859-1')  # Adjust encoding as needed
        print(df.columns)  # Print columns to verify names
        contexts = df['Context'].tolist()  # Use the correct column names
        questions = df['Question'].tolist()
        return [{'context': context, 'question': question} for context, question in zip(contexts, questions)]
    except UnicodeDecodeError as e:
        print(f"Error reading the file: {e}")
        raise
    except KeyError as e:
        print(f"KeyError: {e}. Check the column names in the CSV file.")
        raise

# Function to send a request to the QnA API and measure latency
def send_request(data):
    try:
        start_time = time.time()
        response = requests.post(api_url, json=data)
        end_time = time.time()
        latency = end_time - start_time
        # Print status code and response text for debugging
        print(f"Status Code: {response.status_code}")
        print(f"Response Text: {response.text}")
        if response.status_code == 200:
            try:
                response_json = response.json()  # May raise JSONDecodeError if response is not valid JSON
                return latency, response_json
            except ValueError as e:
                print(f"JSON decoding failed: {e} for response: {response.text}")
                return latency, None
        else:
            print(f"Non-200 status code received: {response.status_code}")
            return latency, None
    except requests.exceptions.RequestException as e:
        print(f"Request failed: {e}")
        return None, None

# Measure latency for multiple concurrent requests
def measure_latency_concurrent_requests(data, num_requests=10):
    latencies = []
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(send_request, data_item) for data_item in data for _ in range(num_requests)]
        for future in concurrent.futures.as_completed(futures):
            latency, _ = future.result()
            if latency is not None:
                latencies.append(latency)
    return latencies

# Update with your dataset file path
file_path = '/content/drive/My Drive/Proplens/qa_data.csv'
sample_data = load_dataset(file_path)

# Run the testing script
latencies = measure_latency_concurrent_requests(sample_data, num_requests=5)
if latencies:
    average_latency = sum(latencies) / len(latencies)
    print(f"Average Latency: {average_latency:.2f} seconds")
    print(f"All Latencies: {latencies}")
else:
    print("No valid latencies recorded.")


Index(['File', 'Context', 'Question', 'Answer'], dtype='object')
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.2242898941040039}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.21341967582702637}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.2206883430480957}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.21271729469299316}
Status Code: 200
Response Text: {"answer":"registration","latency":0.18780279159545898}
Status Code: 200
Response Text: {"answer":"employee credit, post-sales/CRM teams, and sales incentives","latency":0.20363402366638184}
Status Code: 200
Response Text: {"answer":"registration","latency":0.20010781288146973}
Status Code: 200
Response Text: {"answer":"registration","latency":0.19957923889160156

20

In [13]:
import requests
import time
import concurrent.futures
import pandas as pd

# Define the API endpoint (replace with the actual ngrok URL)
api_url = "https://5fb3-34-145-195-49.ngrok-free.app/qna"  # Update with the actual ngrok URL

# Load dataset from CSV
def load_dataset(file_path):
    try:
        df = pd.read_csv(file_path, encoding='ISO-8859-1')  # Adjust encoding as needed
        print(df.columns)  # Print columns to verify names
        contexts = df['Context'].tolist()  # Use the correct column names
        questions = df['Question'].tolist()
        return [{'context': context, 'question': question} for context, question in zip(contexts, questions)]
    except UnicodeDecodeError as e:
        print(f"Error reading the file: {e}")
        raise
    except KeyError as e:
        print(f"KeyError: {e}. Check the column names in the CSV file.")
        raise

# Function to send a request to the QnA API and measure latency
def send_request(data):
    try:
        start_time = time.time()
        response = requests.post(api_url, json=data)
        end_time = time.time()
        latency = end_time - start_time
        # Print status code and response text for debugging
        print(f"Status Code: {response.status_code}")
        print(f"Response Text: {response.text}")
        if response.status_code == 200:
            try:
                response_json = response.json()  # May raise JSONDecodeError if response is not valid JSON
                return latency, response_json
            except ValueError as e:
                print(f"JSON decoding failed: {e} for response: {response.text}")
                return latency, None
        else:
            print(f"Non-200 status code received: {response.status_code}")
            return latency, None
    except requests.exceptions.RequestException as e:
        print(f"Request failed: {e}")
        return None, None

# Measure latency for multiple concurrent requests
def measure_latency_concurrent_requests(data, num_requests=20):
    latencies = []
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(send_request, data_item) for data_item in data for _ in range(num_requests)]
        for future in concurrent.futures.as_completed(futures):
            latency, _ = future.result()
            if latency is not None:
                latencies.append(latency)
    return latencies

# Update with your dataset file path
file_path = '/content/drive/My Drive/Proplens/qa_data.csv'
sample_data = load_dataset(file_path)

# Run the testing script
latencies = measure_latency_concurrent_requests(sample_data, num_requests=5)
if latencies:
    average_latency = sum(latencies) / len(latencies)
    print(f"Average Latency: {average_latency:.2f} seconds")
    print(f"All Latencies: {latencies}")
else:
    print("No valid latencies recorded.")


Index(['File', 'Context', 'Question', 'Answer'], dtype='object')
Status Code: 429
Response Text: <!DOCTYPE html>
<html class="h-full" lang="en-US" dir="ltr">
  <head>
    <link rel="preload" href="https://cdn.ngrok.com/static/fonts/euclid-square/EuclidSquare-Regular-WebS.woff" as="font" type="font/woff" crossorigin="anonymous" />
    <link rel="preload" href="https://cdn.ngrok.com/static/fonts/euclid-square/EuclidSquare-RegularItalic-WebS.woff" as="font" type="font/woff" crossorigin="anonymous" />
    <link rel="preload" href="https://cdn.ngrok.com/static/fonts/euclid-square/EuclidSquare-Medium-WebS.woff" as="font" type="font/woff" crossorigin="anonymous" />
    <link rel="preload" href="https://cdn.ngrok.com/static/fonts/euclid-square/EuclidSquare-Semibold-WebS.woff" as="font" type="font/woff" crossorigin="anonymous" />
    <link rel="preload" href="https://cdn.ngrok.com/static/fonts/euclid-square/EuclidSquare-MediumItalic-WebS.woff" as="font" type="font/woff" crossorigin="anonymous" 

30

In [14]:
import requests
import time
import concurrent.futures
import pandas as pd

# Define the API endpoint (replace with the actual ngrok URL)
api_url = "https://5fb3-34-145-195-49.ngrok-free.app/qna"  # Update with the actual ngrok URL

# Load dataset from CSV
def load_dataset(file_path):
    try:
        df = pd.read_csv(file_path, encoding='ISO-8859-1')  # Adjust encoding as needed
        print(df.columns)  # Print columns to verify names
        contexts = df['Context'].tolist()  # Use the correct column names
        questions = df['Question'].tolist()
        return [{'context': context, 'question': question} for context, question in zip(contexts, questions)]
    except UnicodeDecodeError as e:
        print(f"Error reading the file: {e}")
        raise
    except KeyError as e:
        print(f"KeyError: {e}. Check the column names in the CSV file.")
        raise

# Function to send a request to the QnA API and measure latency
def send_request(data):
    try:
        start_time = time.time()
        response = requests.post(api_url, json=data)
        end_time = time.time()
        latency = end_time - start_time
        # Print status code and response text for debugging
        print(f"Status Code: {response.status_code}")
        print(f"Response Text: {response.text}")
        if response.status_code == 200:
            try:
                response_json = response.json()  # May raise JSONDecodeError if response is not valid JSON
                return latency, response_json
            except ValueError as e:
                print(f"JSON decoding failed: {e} for response: {response.text}")
                return latency, None
        else:
            print(f"Non-200 status code received: {response.status_code}")
            return latency, None
    except requests.exceptions.RequestException as e:
        print(f"Request failed: {e}")
        return None, None

# Measure latency for multiple concurrent requests
def measure_latency_concurrent_requests(data, num_requests=30):
    latencies = []
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(send_request, data_item) for data_item in data for _ in range(num_requests)]
        for future in concurrent.futures.as_completed(futures):
            latency, _ = future.result()
            if latency is not None:
                latencies.append(latency)
    return latencies

# Update with your dataset file path
file_path = '/content/drive/My Drive/Proplens/qa_data.csv'
sample_data = load_dataset(file_path)

# Run the testing script
latencies = measure_latency_concurrent_requests(sample_data, num_requests=5)
if latencies:
    average_latency = sum(latencies) / len(latencies)
    print(f"Average Latency: {average_latency:.2f} seconds")
    print(f"All Latencies: {latencies}")
else:
    print("No valid latencies recorded.")


Streaming output truncated to the last 5000 lines.
    <meta name="author" content="ngrok">
    <meta name="description" content="ngrok is the fastest way to put anything on the internet with a single command.">
    <meta name="robots" content="noindex, nofollow">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link id="style" rel="stylesheet" href="https://cdn.ngrok.com/static/css/error.css">
    <noscript>You have exceeded your limit on connections per minute. This limit will reset within 1 minute. If you expect to continually exceed these limits, please reach out to support (support@ngrok.com) (ERR_NGROK_715)</noscript>
    <script id="script" src="https://cdn.ngrok.com/static/js/error.js" type="text/javascript"></script>
  </head>
  <body class="h-full" id="ngrok">
    <div id="root" data-payload="eyJjZG5CYXNlIjoiaHR0cHM6Ly9jZG4ubmdyb2suY29tLyIsImNvZGUiOiI3MTUiLCJtZXNzYWdlIjoiWW91IGhhdmUgZXhjZWVkZWQgeW91ciBsaW1pdCBvbiBjb25uZWN0aW9ucyBwZXIgbWludXRlLiBUa

Status Code 429: This means "Too Many Requests." It’s a signal from the server that you’re sending too many requests in a given amount of time.
Latency: The average latency for 30 requests is significantly lower (0.41 seconds) compared to 10 requests (0.78 seconds). This might be due to how the server handles burst traffic versus steady traffic.

# RAG **Latency**

In [15]:
import requests
import time
import concurrent.futures
import pandas as pd

# Define the API endpoint (replace with the actual ngrok URL)
api_url = "https://d2f1-34-145-195-49.ngrok-free.app/qna"  # Update with the actual ngrok URL

# Load dataset from CSV
def load_dataset(file_path):
    try:
        df = pd.read_csv(file_path, encoding='ISO-8859-1')  # Adjust encoding as needed
        print(df.columns)  # Print columns to verify names
        contexts = df['Context'].tolist()  # Use the correct column names
        questions = df['Question'].tolist()
        return [{'context': context, 'question': question} for context, question in zip(contexts, questions)]
    except UnicodeDecodeError as e:
        print(f"Error reading the file: {e}")
        raise
    except KeyError as e:
        print(f"KeyError: {e}. Check the column names in the CSV file.")
        raise

# Function to send a request to the QnA API and measure latency
def send_request(data):
    try:
        start_time = time.time()
        response = requests.post(api_url, json=data)
        end_time = time.time()
        latency = end_time - start_time
        # Print status code and response text for debugging
        print(f"Status Code: {response.status_code}")
        print(f"Response Text: {response.text}")
        if response.status_code == 200:
            try:
                response_json = response.json()  # May raise JSONDecodeError if response is not valid JSON
                return latency, response_json
            except ValueError as e:
                print(f"JSON decoding failed: {e} for response: {response.text}")
                return latency, None
        else:
            print(f"Non-200 status code received: {response.status_code}")
            return latency, None
    except requests.exceptions.RequestException as e:
        print(f"Request failed: {e}")
        return None, None

# Measure latency for multiple concurrent requests
def measure_latency_concurrent_requests(data, num_requests=10):
    latencies = []
    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = [executor.submit(send_request, data_item) for data_item in data for _ in range(num_requests)]
        for future in concurrent.futures.as_completed(futures):
            latency, _ = future.result()
            if latency is not None:
                latencies.append(latency)
    return latencies

# Update with your dataset file path
file_path = '/content/drive/My Drive/Proplens/qa_data.csv'
sample_data = load_dataset(file_path)

# Run the testing script
latencies = measure_latency_concurrent_requests(sample_data, num_requests=5)
if latencies:
    average_latency = sum(latencies) / len(latencies)
    print(f"Average Latency: {average_latency:.2f} seconds")
    print(f"All Latencies: {latencies}")
else:
    print("No valid latencies recorded.")


Index(['File', 'Context', 'Question', 'Answer'], dtype='object')
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"Not Found"}
Non-200 status code received: 404
Status Code: 404
Response Text: {"detail":"No